### DISCUSSION TITLE
long-running thread worker doesn't exit when app is quit

### DISCUSSION THREAD

**Discussion by dbieber** [Textualize/textual#3967] [Q&A] [answered]
In the following ExampleApp, I would expect that when the app is quit, the thread worker would terminate too. It does not, and instead the process runs forever.
When I saw that the thread worker does not terminate, my initial thought was perhaps textual failed to set the thread as a daemon, and that this would be an easy fix.

Here's the example code. It continues running even after quitting the app with ctrl-C.

```python
import time
from textual import work
from textual.app import App
from textual.app import ComposeResult
from textual.messages import Message
from textual.widget import Widget
from textual.widgets import Log


class ExampleWidget(Widget):

  class Changed(Message):
    def __init__(self, name: str):
      super().__init__()
      self.name = name

  def on_mount(self):
    self.start_worker()

  @work(thread=True)
  def start_worker(self) -> None:
    while True:
      time.sleep(1)
      self.post_message(self.Changed("And again..."))


class ExampleApp(App):

  def compose(self) -> ComposeResult:
    yield Log()
    yield ExampleWidget()


if __name__ == "__main__":
  app = ExampleApp()
  app.run()
```

---

Surprisingly (to me) if I try to minimize the example further, then it starts working as intended and the bug goes away. In the following code, the app successfully quits on ctrl-C. The difference is that the thread worker is in the App rather than in a widget.

```python
import time
from textual import work
from textual.app import App
from textual.app import ComposeResult
from textual.widgets import Log


class ExampleApp(App):

  def compose(self) -> ComposeResult:
    yield Log()

  def on_mount(self):
    self.start_worker()

  @work(thread=True)
  def start_worker(self):
    log = self.query_one(Log)
    while True:
      log.write_line("And again...")
      time.sleep(1)


if __name__ == "__main__":
  app = ExampleApp()
  app.run()
```

---

I did notice that in https://github.com/Textualize/textual/issues/2593  @davep notes that long-running thread workers are a bad idea. I haven't looked into alternatives yet. If you have a recommendation or rationale, do let me know.

**Comment by davep** [comment, marked as answer]
It might be worth having a glance back over [the guide to workers, and the section on threaded workers](https://textual.textualize.io/guide/workers/#thread-workers) and note the use of `get_current_worker` to get the worker and then checking if the worker `is_cancelled` in the busy loop.

The issue you link to is, if I recall correctly, more about not allowing long-running thread workers to stack up, because eventually you'll have too many threads running.
